File: src/test/java/integration/turn/action/player/ScryXCardsServiceTest.java
package integration.turn.action.player;

import com.matag.cards.Card;
import com.matag.cards.Cards;
import com.matag.game.cardinstance.CardInstance;
import com.matag.game.cardinstance.CardInstanceFactory;
import com.matag.game.status.GameStatus;
import com.matag.game.turn.action.player.ScryXCardsService;
import integration.TestUtils;
import integration.TestUtilsConfiguration;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@ContextConfiguration(classes = {PlayerTestConfiguration.class, TestUtilsConfiguration.class})
public class ScryXCardsServiceTest {
  private int cardInstanceId = 60;

  @Autowired
  private ScryXCardsService scryXCardsService;

  @Autowired
  private TestUtils testUtils;

  @Autowired
  private Cards cards;

  @Autowired
  private CardInstanceFactory cardInstanceFactory;

  @Test
  public void scry1CardWithoutChangingOrder() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    gameStatus.getCurrentPlayer().getLibrary().getCards().clear();

    gameStatus.getCurrentPlayer().getLibrary().addCards(List.of(
      createCardInstance(gameStatus, "Plains"),
      createCardInstance(gameStatus, "Island"),
      createCardInstance(gameStatus, "Swamp"),
      createCardInstance(gameStatus, "Mountain"),
      createCardInstance(gameStatus, "Forest")
    ));

    // When
    scryXCardsService.scryXCards(gameStatus, List.of(1));

    // Then
    assertCards(gameStatus, "Plains", "Island", "Swamp", "Mountain", "Forest");
  }

  @Test
  public void scry1CardPutToBottom() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    gameStatus.getCurrentPlayer().getLibrary().getCards().clear();

    gameStatus.getCurrentPlayer().getLibrary().addCards(List.of(
      createCardInstance(gameStatus, "Plains"),
      createCardInstance(gameStatus, "Island"),
      createCardInstance(gameStatus, "Swamp"),
      createCardInstance(gameStatus, "Mountain"),
      createCardInstance(gameStatus, "Forest")
    ));

    // When
    scryXCardsService.scryXCards(gameStatus, List.of(-1));

    // Then
    assertCards(gameStatus, "Island", "Swamp", "Mountain", "Forest", "Plains");
  }

  @Test
  public void scry4Cards() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    gameStatus.getCurrentPlayer().getLibrary().getCards().clear();

    gameStatus.getCurrentPlayer().getLibrary().addCards(List.of(
      createCardInstance(gameStatus, "Plains"),
      createCardInstance(gameStatus, "Island"),
      createCardInstance(gameStatus, "Swamp"),
      createCardInstance(gameStatus, "Mountain"),
      createCardInstance(gameStatus, "Forest")
    ));

    // When
    scryXCardsService.scryXCards(gameStatus, List.of(2, -1, -2, 1));

    // Then
    assertCards(gameStatus, "Mountain", "Plains", "Forest", "Island", "Swamp");
  }

  private CardInstance createCardInstance(GameStatus gameStatus, String cardName) {
    var card = cards.get(cardName);
    var cardInstance = cardInstanceFactory.create(gameStatus, ++cardInstanceId, card, "player-name", "player-name");
    gameStatus.getActivePlayer().getBattlefield().addCard(cardInstance);
    return cardInstance;
  }

  private void assertCards(GameStatus gameStatus, String... cardNames) {
    var actualCardNames = gameStatus.getCurrentPlayer().getLibrary().getCards().stream()
      .map(CardInstance::getCard).map(Card::getName)
      .collect(Collectors.toList());

    assertThat(actualCardNames).containsExactly(cardNames);
  }
}

File: src/test/kotlin/integration/turn/action/player/ScryXCardsServiceTest.kt
package integration.turn.action.player

import com.matag.cards.Cards
import com.matag.game.MatagGameApplication
import com.matag.game.cardinstance.CardInstance
import com.matag.game.cardinstance.CardInstanceFactory
import com.matag.game.status.GameStatus
import com.matag.game.turn.action.player.ScryXCardsService
import integration.TestUtils
import integration.TestUtilsConfiguration
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles

@SpringBootTest(classes = [MatagGameApplication::class], webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("test")
@Import(TestUtilsConfiguration::class)
class ScryXCardsServiceTest {
    private var cardInstanceId = 60

    @Autowired
    private val scryXCardsService: ScryXCardsService? = null

    @Autowired
    private val testUtils: TestUtils? = null

    @Autowired
    private val cards: Cards? = null

    @Autowired
    private val cardInstanceFactory: CardInstanceFactory? = null

    @Test
    fun scry1CardWithoutChangingOrder() {
        // Given
        val gameStatus = testUtils!!.testGameStatus()
        gameStatus.currentPlayer.library.cards = listOf()

        gameStatus.currentPlayer.library.addCards(
            listOf(
                createCardInstance(gameStatus, "Plains"),
                createCardInstance(gameStatus, "Island"),
                createCardInstance(gameStatus, "Swamp"),
                createCardInstance(gameStatus, "Mountain"),
                createCardInstance(gameStatus, "Forest")
            )
        )

        // When
        scryXCardsService!!.scryXCards(gameStatus, listOf(1))

        // Then
        assertCards(gameStatus, "Plains", "Island", "Swamp", "Mountain", "Forest")
    }

    @Test
    fun scry1CardPutToBottom() {
        // Given
        val gameStatus = testUtils!!.testGameStatus()
        gameStatus.currentPlayer.library.cards = listOf()

        gameStatus.currentPlayer.library.addCards(
            listOf(
                createCardInstance(gameStatus, "Plains"),
                createCardInstance(gameStatus, "Island"),
                createCardInstance(gameStatus, "Swamp"),
                createCardInstance(gameStatus, "Mountain"),
                createCardInstance(gameStatus, "Forest")
            )
        )

        // When
        scryXCardsService!!.scryXCards(gameStatus, listOf(-1))

        // Then
        assertCards(gameStatus, "Island", "Swamp", "Mountain", "Forest", "Plains")
    }

    @Test
    fun scry4Cards() {
        // Given
        val gameStatus = testUtils!!.testGameStatus()
        gameStatus.currentPlayer.library.cards = listOf()

        gameStatus.currentPlayer.library.addCards(
            listOf(
                createCardInstance(gameStatus, "Plains"),
                createCardInstance(gameStatus, "Island"),
                createCardInstance(gameStatus, "Swamp"),
                createCardInstance(gameStatus, "Mountain"),
                createCardInstance(gameStatus, "Forest")
            )
        )

        // When
        scryXCardsService!!.scryXCards(gameStatus, listOf(2, -1, -2, 1))

        // Then
        assertCards(gameStatus, "Mountain", "Plains", "Forest", "Island", "Swamp")
    }

    private fun createCardInstance(gameStatus: GameStatus, cardName: String): CardInstance {
        val card = cards!!.get(cardName)
        val cardInstance =
            cardInstanceFactory!!.create(gameStatus, ++cardInstanceId, card, "player-name", "player-name")
        gameStatus.activePlayer.battlefield.addCard(cardInstance)
        return cardInstance
    }

    private fun assertCards(gameStatus: GameStatus, vararg cardNames: String?) {
        val actualCardNames = gameStatus.currentPlayer.library.cards
            .map { it.card?.name }

        Assertions.assertThat(actualCardNames).containsExactly(*cardNames)
    }
}